File: 백준/Gold/1753. 최단경로/최단경로.js
const fs = require("fs");

const stdin = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [v] = stdin[0].split(" ").map((num) => parseInt(num));
const start = parseInt(stdin[1]);

// [weight, vertex]
const adj = Array.from({ length: v + 1 }, () => new Array());

stdin.slice(2).forEach((row) => {
  const [u, v, w] = row.split(" ").map((num) => parseInt(num));
  adj[u].push([w, v]);
});

const weight = 0;
const vertex = 1;
const INF = 1e9;

class PriorityQueue {
  constructor() {
    this.heap = [];
    this.size = 0;
  }

  push(x) {
    this.heap[++this.size] = x;
    let index = this.size;

    while (index != 1) {
      let parent = parseInt(index / 2);

      if (this.heap[index][weight] >= this.heap[parent][weight]) break;

      [this.heap[index], this.heap[parent]] = [
        this.heap[parent],
        this.heap[index],
      ];

      index = parent;
    }
  }

  top() {
    return this.heap[1];
  }

  pop() {
    if (this.empty()) return;
    this.heap[1] = this.heap[this.size--];
    let index = 1;

    while (2 * index <= this.size) {
      let [leftChildren, rightChildren] = [index * 2, index * 2 + 1];
      let minChildren = leftChildren;

      if (
        rightChildren <= this.size &&
        this.heap[rightChildren][weight] < this.heap[leftChildren][weight]
      ) {
        minChildren = rightChildren;
      }

      if (this.heap[index][weight] <= this.heap[minChildren][weight]) break;

      [this.heap[index], this.heap[minChildren]] = [
        this.heap[minChildren],
        this.heap[index],
      ];

      index = minChildren;
    }
  }

  empty() {
    return this.size === 0;
  }
}

const djik = (v, start) => {
  const dist = Array.from({ length: v + 1 }, () => INF);
  const pq = new PriorityQueue();

  dist[start] = 0;

  pq.push([dist[start], start]);

  while (!pq.empty()) {
    let [currentWeight, currentVertex] = pq.top();

    pq.pop();

    if (dist[currentVertex] !== currentWeight) continue;

    for (let next of adj[currentVertex]) {
      const [nextWeight, nextVertex] = next;

      if (dist[nextVertex] > dist[currentVertex] + nextWeight) {
        dist[nextVertex] = dist[currentVertex] + nextWeight;
        pq.push([dist[nextVertex], nextVertex]);
      }
    }
  }

  console.log(
    dist
      .slice(1)
      .map((value) => (value === INF ? "INF" : value))
      .join("\n")
  );
};

djik(v, start);

File: 백준/Gold/1753. 최단경로/최단경로.cc
#include <bits/stdc++.h>

#define X first
#define Y second

using namespace std;

// {비용, 정점 번호}
vector<pair<int, int>> adj[20005];
const int INF = 1e9 + 10;
int d[20005];  // 최단거리 테이블

void djik(int v, int st) {
  // 전체 테이블을 무한대로 초기화
  fill(d, d + v + 1, INF);

  // 최소 힙
  priority_queue<pair<int, int>, vector<pair<int, int>>,
                 greater<pair<int, int>>>
      pq;

  // 시작점의 거리는 0이다.
  d[st] = 0;

  // 우선순위 큐에 시작점을 넣는다.
  pq.push({d[st], st});

  while (!pq.empty()) {
    auto cur = pq.top();
    pq.pop();

    // 우선순위 큐에 있는 점이 올바른 값인지 확인한다.
    if (d[cur.Y] != cur.X) continue;

    // 올바르다면 현재 정점에서 갈 수있는 거리와 테이블의 최단 거리를 비교한다.
    for (auto next : adj[cur.Y]) {
      if (d[next.Y] > d[cur.Y] + next.X) {
        d[next.Y] = d[cur.Y] + next.X;
        pq.push({d[next.Y], next.Y});
      }
    }
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int v, e, st;

  cin >> v >> e >> st;

  for (int i = 0; i < e; i++) {
    int u, v, w;
    cin >> u >> v >> w;

    adj[u].push_back({w, v});
  }

  djik(v, st);

  for (int i = 1; i <= v; i++) {
    if (d[i] == INF)
      cout << "INF\n";
    else
      cout << d[i] << '\n';
  }

  return 0;
}
